app/src/main/java/org/grammatek/simacorrect/spellcheckerservice/YfirlesturAnnotation.kt

```diff
@@ -35,105 +40,109 @@ class YfirlesturAnnotation(
-        // are inside of multi word annotations.
-        val annotationsList = _annotations.groupBy { it.toKey() }
+    fun getSuggestionsForAnnotatedWords(suggestionsLimit: Int, dict: ArrayList<String>): List<SuggestionsInfo> {
         val suggestionList = mutableListOf<SuggestionsInfo>()
+        for (i in _annotations.indices) {
+            // Group annotations in a list that have the same start AND end index.
+            // Necessary to distinguish between cases where single word annotations
+            // are inside of multi word annotations.
+            val sentenceAnnotations = _annotations[i].groupBy { it.toKey() }
+            for((_, tokenAnnotations) in sentenceAnnotations) {
+                val suggestions = mutableListOf<String>()
+                val flags = mutableListOf<Int>()
+                var startChar = 0
+                var endChar = 0
 
-        for((_, annotations) in annotationsList) {
-            val suggestions = mutableListOf<String>()
-            var flag = 0
-            var startChar = 0
-            var endChar = 0
-            var sequence = 0
-            for (annotation in annotations) {
-                if(annotation.code == null) {
-                    continue
-                }
-                sequence = getSequence(annotation.startChar!!, annotation.endChar!!)
-                flag = determineSuggestionFlag(annotation.code.toString())
-
-                val suggestion: String = correctSuggestion(
-                    annotation.suggest ?: "", annotation.end!!-annotation.start!!
-                )
-
-                // Take into account that Yfirlestur includes whitespaces in their annotation
-                startChar = if (annotation.startChar!! != 0) {
-                    annotation.startChar!! + 1
-                } else {
-                    annotation.startChar!!
+                for (annotation in tokenAnnotations) {
+                    val startCharIndexedForUnalteredText = tokensIndices[i][annotation.start!!].startChar
+                    val endCharIndexedForUnalteredText = tokensIndices[i][annotation.end!!].endChar
+                    val word = _unalteredOriginalText!!.substring(startCharIndexedForUnalteredText, endCharIndexedForUnalteredText + 1)
+                    if(annotation.code == null || dict.contains(word)) {
+                        continue
+                    }
+
+                    val flag = determineSuggestionFlag(annotation.code.toString())
+                    if(flag == 0) {
+                        continue
+                    }
+                    flags.add(flag)
+
+                    var suggestion = annotation.suggest
+                    val annotationTokenStartIndex = annotation.start!!
+                    var annotationTokenEndIndex = annotation.end!!
+                    // if words are split up we have to fix the indexing from Yfirlestur
+                    // the method called tries to determine if that is the case.
```

Comment:
add an example

Reply:
Interestingly enough, this method `adjustTokenIndices` doesn't appear to be necessary anymore.